win32ss/user/ntuser/csr.c

```diff
@@ -80,17 +80,16 @@ CsrClientCallServer(IN OUT PCSR_API_MESSAGE ApiMessage,
     ULONG PointerCount;
     PULONG_PTR OffsetPointer;
 #endif
-
+    
     /* Do we have a connection to CSR yet? */
     if (!CsrApiPort)
         return STATUS_INVALID_PORT_HANDLE;
 
     /* Fill out the Port Message Header */
     ApiMessage->Header.u2.ZeroInit = 0;
-    ApiMessage->Header.u1.s1.TotalLength = DataLength +
-        sizeof(CSR_API_MESSAGE) - sizeof(ApiMessage->Data); // FIELD_OFFSET(CSR_API_MESSAGE, Data) + DataLength;
-    ApiMessage->Header.u1.s1.DataLength = DataLength +
-        FIELD_OFFSET(CSR_API_MESSAGE, Data) - sizeof(ApiMessage->Header); // ApiMessage->Header.u1.s1.TotalLength - sizeof(PORT_MESSAGE);
+    ApiMessage->Header.u1.s1.TotalLength = FIELD_OFFSET(CSR_API_MESSAGE, Data) + DataLength;
+    ApiMessage->Header.u1.s1.DataLength = ApiMessage->Header.u1.s1.TotalLength - 
+        sizeof(ApiMessage->Header);
```

Comment:
So I understand the old commented code was more correct than the one that was enabled?
Also, your `sizeof(ApiMessage->Header)` seems to have bad indent.

Reply:
They're equivalent. sizeof(struct) - sizeof(last member) is just more convoluted, so the FIELD_OFFSET is better.

Reply:
Well, adding CHAR Buffer[1] or whatever to CSR_API_MESSAGE would certainly help code readability here and in other places .

Reply:
Exactly what the "Data" member is for.

Reply:
Not exactly what I meant. You can't use FIELD_OFFSET(x, y[foo]) if y is an union. See https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/reactos/subsys/csr/csrmsg.h;hb=282b4397e03cd4f71c55c06b84f2b63b54f6dd38#l139

Reply:
Ok ok. In that particular case of the CSR_API_MESSAGE the code is readable & understandable (and we are not trying to find field offsets of some member from inside the union, so everything is fine).